zrml/court/src/lib.rs

```diff
@@ -190,8 +208,130 @@ mod pallet {
         }
 
         // Retrieves a juror from the storage
-        fn juror(account_id: &T::AccountId) -> Result<Juror<BalanceOf<T>>, DispatchError> {
-            Jurors::<T>::get(account_id).ok_or_else(|| Error::<T>::JurorDoesNotExists.into())
+        fn mutate_juror<F>(account_id: &T::AccountId, mut cb: F) -> DispatchResult
+        where
+            F: FnMut(&mut Juror<BalanceOf<T>>) -> DispatchResult,
+        {
+            Jurors::<T>::try_mutate(account_id, |opt| {
+                if let Some(el) = opt {
+                    cb(el)?;
+                } else {
+                    return Err(Error::<T>::JurorDoesNotExists.into());
+                }
+                Ok(())
+            })
+        }
+
+        // Slashes 20% of staked funds and removes them from the list of jurors
+        fn punish_tardy_jurors(
+            requested_jurors: &mut Vec<(T::AccountId, T::BlockNumber)>,
+        ) -> DispatchResult {
+            let mut i = 0;
+            while i < requested_jurors.len() {
+                let should_remove = {
+                    // `get` will never panic
+                    let (ai, _) = requested_jurors.get(i).unwrap();
+                    let juror = Self::juror(ai)?;
+                    matches!(juror.status, JurorStatus::Tardy)
+                };
+                if should_remove {
+                    // `remove` will never panic
+                    let (ai, _) = requested_jurors.remove(i);
+                    let reserved = CurrencyOf::<T>::reserved_balance(&ai);
+                    // Division will never overflow
+                    let slash = reserved / BalanceOf::<T>::from(5u8);
+                    CurrencyOf::<T>::repatriate_reserved(
+                        &ai,
+                        &T::TreasuryAccount::get(),
+                        slash,
+                        BalanceStatus::Free,
+                    )?;
+                    Jurors::<T>::remove(ai);
```

Comment:
>The account that was removed from the court has no chance anymore to unreserve it's reserved funds, since the exit_court dispatchable, which is the only dispatchable that unreserves the reserved currency of a juror, will result in an error, because the juror in question cannot be found anymore.

IIRC all unreserves were handled on https://github.com/zeitgeistpm/zeitgeist/pull/276 but yeah, this should be handled in this PR. Fixed.

> What happens with other votes the juror submitted? Let's assume the juror is part of 4 court cases. In the first case (1) the juror is marked as Tardy. In the second case the juror misbehaves again, therefore the juror is removed from the court. However, the juror might have already voted for case 3 and case 4.

Great finding, this is indeed a big problem!
If possible, I would like to tackle this issue in a following PR.

> This brings me to another question: The jurors who voted for a losing outcome and get 50% of their bond slashed are still part of the court. Does this mean that during the next case, they will have only 50% of their initial bond on stake? And the fifth time they lose (without belonging to the losing majority), it's only 0.5.pow(5) = 0.03125 of the initial bond that's on stake? I think we have to discuss this corner case.

I think this situation is OK. If an account loses a case, then he can exit the court anytime.
It is also possible to impose a hard cap for consecutive losses

Reply:
About the last point: I was worried about the circumstance that a juror who has lost a majority of his stake can still cast votes without much to lose. I also think that this should be no problem, because somebody who casts malicious votes will eventually be flagged as tardy and removed.

Reply:
Hum... Indeed...

What are your thoughts, @lsaether ?

Reply:
I will try to tackle this problem on https://github.com/zeitgeistpm/zeitgeist/pull/276